From: Balbir Singh <balbirs@nvidia.com>
Subject: mm-huge_memoryc-introduce-folio_split_unmapped-v2
Date: Sat, 15 Nov 2025 19:40:41 +1100

- Fix a bug reported by kernel-test-robot, nr_shmem_dropped needs to
  be updated and returned from the helper function

- Add a debug warning folio_split_unmapped() should be called only on
  anonymous folios

Link: https://lkml.kernel.org/r/20251115084041.3914728-1-balbirs@nvidia.com
Signed-off-by: Balbir Singh <balbirs@nvidia.com>
Suggested-by: Zi Yan <ziy@nvidia.com>
Acked-by: Zi Yan <ziy@nvidia.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: Joshua Hahn <joshua.hahnjy@gmail.com>
Cc: Rakie Kim <rakie.kim@sk.com>
Cc: Byungchul Park <byungchul@sk.com>
Cc: Gregory Price <gourry@gourry.net>
Cc: Ying Huang <ying.huang@linux.alibaba.com>
Cc: Alistair Popple <apopple@nvidia.com>
Cc: Oscar Salvador <osalvador@suse.de>
Cc: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Cc: Baolin Wang <baolin.wang@linux.alibaba.com>
Cc: "Liam R. Howlett" <Liam.Howlett@oracle.com>
Cc: Nico Pache <npache@redhat.com>
Cc: Ryan Roberts <ryan.roberts@arm.com>
Cc: Dev Jain <dev.jain@arm.com>
Cc: Barry Song <baohua@kernel.org>
Cc: Lyude Paul <lyude@redhat.com>
Cc: Danilo Krummrich <dakr@kernel.org>
Cc: David Airlie <airlied@gmail.com>
Cc: Simona Vetter <simona@ffwll.ch>
Cc: Ralph Campbell <rcampbell@nvidia.com>
Cc: Mika Penttilä <mpenttil@redhat.com>
Cc: Matthew Brost <matthew.brost@intel.com>
Cc: Francois Dugast <francois.dugast@intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/huge_memory.c |   16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

--- a/mm/huge_memory.c~mm-huge_memoryc-introduce-folio_split_unmapped-v2
+++ a/mm/huge_memory.c
@@ -3743,12 +3743,11 @@ static int __folio_freeze_and_split_unma
 					     struct page *split_at, struct xa_state *xas,
 					     struct address_space *mapping, bool do_lru,
 					     struct list_head *list, enum split_type split_type,
-					     pgoff_t end, int extra_pins)
+					     pgoff_t end, int *nr_shmem_dropped, int extra_pins)
 {
 	struct folio *end_folio = folio_next(folio);
 	struct folio *new_folio, *next;
 	int old_order = folio_order(folio);
-	int nr_shmem_dropped = 0;
 	int ret = 0;
 	struct deferred_split *ds_queue;
 
@@ -3851,9 +3850,10 @@ static int __folio_freeze_and_split_unma
 				continue;
 			}
 
+			VM_WARN_ON_ONCE(!nr_shmem_dropped);
 			/* Drop folio beyond EOF: ->index >= end */
-			if (shmem_mapping(mapping))
-				nr_shmem_dropped += nr_pages;
+			if (shmem_mapping(mapping) && nr_shmem_dropped)
+				*nr_shmem_dropped += nr_pages;
 			else if (folio_test_clear_dirty(new_folio))
 				folio_account_cleaned(
 					new_folio, inode_to_wb(mapping->host));
@@ -4039,7 +4039,8 @@ static int __folio_split(struct folio *f
 	}
 
 	ret = __folio_freeze_and_split_unmapped(folio, new_order, split_at, &xas, mapping,
-						true, list, split_type, end, extra_pins);
+						true, list, split_type, end, &nr_shmem_dropped,
+						extra_pins);
 fail:
 	if (mapping)
 		xas_unlock(&xas);
@@ -4104,9 +4105,10 @@ int folio_split_unmapped(struct folio *f
 {
 	int extra_pins, ret = 0;
 
-	VM_WARN_ON_FOLIO(folio_mapped(folio), folio);
+	VM_WARN_ON_ONCE_FOLIO(folio_mapped(folio), folio);
 	VM_WARN_ON_ONCE_FOLIO(!folio_test_locked(folio), folio);
 	VM_WARN_ON_ONCE_FOLIO(!folio_test_large(folio), folio);
+	VM_WARN_ON_ONCE_FOLIO(!folio_test_anon(folio), folio);
 
 	if (!can_split_folio(folio, 1, &extra_pins))
 		return -EAGAIN;
@@ -4114,7 +4116,7 @@ int folio_split_unmapped(struct folio *f
 	local_irq_disable();
 	ret = __folio_freeze_and_split_unmapped(folio, new_order, &folio->page, NULL,
 						NULL, false, NULL, SPLIT_TYPE_UNIFORM,
-						0, extra_pins);
+						0, NULL, extra_pins);
 	local_irq_enable();
 	return ret;
 }
_
